### PR TITLE
AJ-1625: swagger UI for control plane vs data plane

### PR DESCRIPTION
**Pull request by davidangb**
* add a new `cwds-api-docs.yaml` OpenAPI spec which contains only the cWDS-enabled endpoints
* move `…resources/static/swagger/swagger-ui.html` to `…resources/swagger-ui-data-plane.html`, so it can't be served as a static resource anymore
* add `…resources/swagger-ui-control-plane.html`, with slightly different config than the data plane html
* add a new `SwaggerUiConfig` which reads the appropriate html file based on control-plane vs data-plane profile and stashes the html into a String
* add a new `SwaggerUiController` RestController which responds at /swagger/swagger-ui.html and responds with the stashed html
* add redirects from `/`, `/swagger`, and `/swagger/` to /swagger/swagger-ui.html for convenience

### CODE REVIEW

File: service/src/main/java/org/databiosphere/workspacedataservice/controller/SwaggerUiController.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+    String content;
+
+    if (environment.matchesProfiles("control-plane")) {
+      // replace the title
+      content =
+          template.replace("${TITLE}", "cWDS API").replace("${APISPEC}", "cwds-api-docs.yaml");
+    } else {
+      content = template.replace("${TITLE}", "WDS API").replace("${APISPEC}", "openapi-docs.yaml");
+    }
```
Comment (davidangb) [inline service/src/main/java/org/databiosphere/workspacedataservice/controller/SwaggerUiController.java:47]: I thought about moving these strings into a config file (application-control-plane.yml, application-data-plane.yml), but … it felt like overkill.

Another option is to have two separate .html files, pre-populated with the correct values, and serve the appropriate one based on control-plane vs data-plane profiles. That would be more performant … but this does not feel like a particularly performance-sensitive spot.

File: service/src/main/resources/static/swagger/cwds-api-docs.yaml [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  /job/v1/instance/{instanceUuid}:
+    $ref: 'apis-v1.yaml#/paths/~1job~1v1~1instance~1{instanceUuid}'
+
+  ##############################
+  # General WDS Information APIs
+  ##############################
+  /status:
+    $ref: 'openapi-docs.yaml#/paths/~1status'
+  /version:
+    $ref: 'openapi-docs.yaml#/paths/~1version'
```
Comment (davidangb) [inline service/src/main/resources/static/swagger/cwds-api-docs.yaml:53]: as discussed over in https://github.com/DataBiosphere/terra-workspace-data-service/pull/572#discussion_r1501213706, we may want the Capabilities API at some point; we can add that later if we want it.

File: service/src/main/resources/swagger-ui-control-plane.html [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+      plugins: [
+        SwaggerUIBundle.plugins.DownloadUrl,
+        cleanupPlugin
+      ],
+      layout: 'StandaloneLayout',
+      displayOperationId: true,       // show generated-client method names for each endpoint
+      defaultModelsExpandDepth: -1,   // hide the huge list of schemas under the routes
+      defaultModelRendering: 'model', // schema has the descriptions, unlike the example value
+      defaultModelExpandDepth: 2,     // affects the schema shown for a request or response
+      displayRequestDuration: true,   // show timing
```
Comment (davidangb) [inline service/src/main/resources/swagger-ui-control-plane.html:57]: compared to the data plane HTML, this is missing: `docExpansion: 'none', filter: true`. Those features seemed not useful with such a small set of APIs in the control plane.

File: service/src/main/java/org/databiosphere/workspacedataservice/controller/SwaggerUiConfig.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+  @DataPlane
+  @Bean(name = "swaggerHtml")
+  public String dataPlaneSwaggerHtml(
+      @Value("classpath:swagger-ui-data-plane.html") Resource swaggerUiDataPlaneResource)
+      throws IOException {
+    return StreamUtils.copyToString(
+        swaggerUiDataPlaneResource.getInputStream(), StandardCharsets.UTF_8);
+  }
+}
```
Comment (davidangb) [inline service/src/main/java/org/databiosphere/workspacedataservice/controller/SwaggerUiConfig.java:40]: I also considered having a single HTML file containing markers like `${TITLE}`, then performing string-replacement inside the controller. But when I got to the point of setting different flags in the swagger-ui javascript (e.g. removing `docExpansion: 'none', filter: true`), that got unweildy.

File: service/src/main/java/org/databiosphere/workspacedataservice/controller/SwaggerUiController.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+public class SwaggerUiController {
+  /**
+   * holds the HTML content of the swagger-ui.
+   *
+   * @see SwaggerUiConfig
+   */
+  final String swaggerUiContent;
+
+  public SwaggerUiController(@Qualifier("swaggerHtml") String swaggerUiContent) {
+    this.swaggerUiContent = swaggerUiContent;
```
Comment (ashanhol) [inline service/src/main/java/org/databiosphere/workspacedataservice/controller/SwaggerUiController.java:25]: I think I'm missing something Spring Magic related. SwaggerUiConfig returns data or controller depending on the annotation but how does that hook up here?

File: service/src/main/resources/static/swagger/cwds-api-docs.yaml [hunk 6]
```diff
@@ -0,0 +1,50 @@
+openapi: 3.0.3
+
+info:
+  title: cWDS API Spec
+  version: v1
+  description: |
+    cWDS - the (c)control-plane (W)orkspace (D)ata (S)ervice - is responsible for importing
```
[the remark this answers is not in the document]
Comment (davidangb) [reply service/src/main/resources/static/swagger/cwds-api-docs.yaml]: fixed :)

File: service/src/main/java/org/databiosphere/workspacedataservice/controller/SwaggerUiController.java (hunk 5 above)
Comment (davidangb) [reply service/src/main/java/org/databiosphere/workspacedataservice/controller/SwaggerUiController.java:25]: `SwaggerUiConfig` produces a bean of type String. The bean's name is "swaggerHtml", as specified by:
```
@Bean(name = "swaggerHtml")
```

Here, Spring uses dependency injection to supply the String argument to this constructor. The `@Qualifier("swaggerHtml")` tells it to look for a bean named "swaggerHtml".

**Review by calypsomatic — APPROVED** [review]

File: service/src/main/java/org/databiosphere/workspacedataservice/controller/SwaggerUiConfig.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+  @Bean(name = "swaggerHtml")
+  public String controlPlaneSwaggerHtml(
+      @Value("classpath:swagger-ui-control-plane.html") Resource swaggerUiControlPlaneResource)
+      throws IOException {
+    return StreamUtils.copyToString(
+        swaggerUiControlPlaneResource.getInputStream(), StandardCharsets.UTF_8);
+  }
+
+  @DataPlane
+  @Bean(name = "swaggerHtml")
```
Comment (jladieu) [inline service/src/main/java/org/databiosphere/workspacedataservice/controller/SwaggerUiConfig.java:33]: ahh neat, I didn't know you could use `@Qualifier` with `@Bean(name=...)` -- I thought you had to use `@Qualifier` in both places!

File: service/src/main/resources/swagger-ui-control-plane.html (hunk 3 above)
Comment (jladieu) [reply service/src/main/resources/swagger-ui-control-plane.html:57]: Would you consider listing them in here with their default values, and commenting them with the same explanation inline?  I'm thinking of at some point doing a visual diff and it'd be helpful to aid knowing what configs are different and by what values.

**Review by jladieu — APPROVED** [review]
This looks great!  One optional suggestion about explicitly calling out the configs that are different, but LGTM either way.

File: service/src/main/resources/swagger-ui-control-plane.html (hunk 3 above)
Comment (davidangb) [reply service/src/main/resources/swagger-ui-control-plane.html:57]: added these back in!

File: service/src/main/java/org/databiosphere/workspacedataservice/controller/SwaggerUiController.java (hunk 5 above)
Comment (ashanhol) [reply service/src/main/java/org/databiosphere/workspacedataservice/controller/SwaggerUiController.java:25]: oh wow thats really cool!

**Review by ashanhol — APPROVED** [review]
It's cool to see a UI spring example

**Pull request merged**